coordinator.lic

```diff
@@ -0,0 +1,478 @@
+    pause_script('sanowret-crystal') if task[:pause_sanowret]
+    if task[:walk_to]
+      case task[:walk_to]
+      when 'safe_room'
+        walk_to(safe_room)
+      else
+        walk_to(task[:walk_to])
+      end
+    end
+    if task[:buff]
+      wait_for_script_to_complete('buff', [task[:buff]])
+    end
+    if task[:town] && DRStats.necromancer?
+      wait_for_script_to_complete('release-necro')
+    end
+    if action = task[:action]
+      unless all_actions.include? action
+        raise ArgumentError, "Action not in whitelist: #{action}"
+      end
+      send(action, task)
+    elsif script = task[:script]
+      play_song?(false) if task[:play_song]
+      wait_for_script_to_complete(script, task[:args])
+      stop_play if task[:play_song]
+    else
+      raise ArgumentError, "Expected action or script: #{task}"
+    end
+    unpause_script('sanowret-crystal') if task[:pause_sanowret]    
+  end
+  
+  def run_hunting_task(task)
+    file = task[:file]
+    echo("Starting hunting-buddy with file #{file}") if @debug
+    pause_script('sanowret-crystal') if task[:pause_sanowret]
+    start_script('hunting-buddy', [file])
+    pause 5
+    next_check = Time.now + 60
+    while Script.running?('hunting-buddy')
+      if Time.now > next_check
+        $HUNTING_BUDDY.stop_hunting if (task[:stop_on] && evaluate_predicate?(task[:stop_on]))
+        next_check = Time.now + 60
+      end
+      pause 5
+    end
+    update_box_count
+    safe_room = @settings.safe_room
+    if task[:safe_room]
+      safe_room = task[:safe_room]['id']
+    end
+    walk_to(safe_room)
+    unpause_script('sanowret-crystal') if task[:pause_sanowret]
+  end
+  
+  #
+  # Predicates
+  #  
+
+  def all_predicates
+    [
+      'hunt_done?',
+      'run_crossing_training?',
+      'pet_boxes?'
+    ]
+  end
+
+  def hunt_done?
+    # TODO: would be nice to not have to double-check predicates here.
+    !hunting_task
+  end
+  
+  def run_crossing_training?
+    @settings.crossing_training.any? { |skill| DRSkill.getxp(skill) < 28 }
+  end
+
+  def pet_boxes?
+    @num_pet_boxes > 0
+  end
+  
+  #
+  # Actions
+  #
+  def all_actions
+    [
+      'pick_boxes',
+      'train_performance',
+      'train_outfitting',
+      'train_engineering',
+      'train_forging',
+      'crossing_training',
+      'infuse_osrel',
+      'check_favors',
+    ]
+  end
+
+  def pick_boxes(task)
+    walk_to(@settings.lockpick_room_id) if @settings.lockpick_room_id
+    if @settings.waggle_sets['locksmithing']
+      wait_for_script_to_complete('buff', ['locksmithing'])
+    end
+    wait_for_script_to_complete('pick')
+    update_box_count
+  end
+
+  def infuse_osrel(task)
+    if @settings.osrel_amount && DRSpells.active_spells['Osrel Meraud']
+      infuse_om(!@settings.osrel_no_harness, @settings.osrel_amount)
+    else
+      echo '***OM IS DOWN OR NO osrel_amount***'
+      task[:deleted] = true
+    end
+  end
+
+  def check_favors(task)
```

Comment:
💡 A lot of this function looks to be identical to `training-manager`, a good candidate for a common file.